pvlib/modelchain.py

```diff
@@ -810,32 +810,101 @@ def run_model(self, times=None, weather=None):
-            If None, assumes air temperature is 20 C, wind speed is 0
-            m/s and irradiation calculated from clear sky data. Column
-            names must be 'wind_speed', 'temp_air', 'dni', 'ghi', 'dhi'.
-            Do not pass incomplete irradiation data. Use method
-            :py:meth:`~pvlib.modelchain.ModelChain.complete_irradiance`
-            instead.
+            If None, assumes air temperature is 20 C, wind speed is 0 m/s and
+            irradiation calculated from clear sky data.
+            For conventional weather data, column names must be 'wind_speed',
+            'temp_air', 'dni', 'ghi', 'dhi'. Do not pass incomplete irradiation
+            data, instead use method
+            :py:meth:`~pvlib.modelchain.ModelChain.complete_irradiance`.
+            For complete PV reference device data, column names must be 'F' and
+            'H', for effective irradiance ratio and effective temperature
+            ratio, respectively. See :py:func:`~pvlib.pvsystem.sdm_campanelli`.
+            If either 'F' or 'H' is missing, then convential weather data are
+            also required to compute the missing 'F' or 'H'.
 
         Returns
         -------
         self
 
-        Assigns attributes: times, solar_position, airmass, irradiance,
-        total_irrad, effective_irradiance, weather, temps, aoi,
-        aoi_modifier, spectral_modifier, dc, ac, losses.
+        For conventional weather data, assigns attributes: times,
+        solar_position, airmass, irradiance, total_irrad, effective_irradiance,
+        weather, temps, aoi, aoi_modifier, spectral_modifier, dc, ac, losses.
+        For complete reference device data, only assigns attributes: times, dc,
+        ac, losses.
         """
 
-        self.prepare_inputs(times, weather)
-        self.aoi_model()
-        self.spectral_model()
-        self.effective_irradiance_model()
-        self.temp_model()
+        if weather is not None and 'F' in weather:
+            # Use PV reference device weather information for F
+            self.F = weather['F']
+        else:
+            # Clear any stale value
+            if hasattr(self, 'F'):
+                print('F cleared')
+                del self.F
+        if weather is not None and 'H' in weather:
+            # Use PV reference device weather information for H
+            self.H = weather['H']
+        else:
+            # Clear any stale value
+            if hasattr(self, 'H'):
+                print('H cleared')
+                del self.H
+        if not hasattr(self, 'F') and not hasattr(self, 'H'):
+            # Use conventional (meterological) weather information
+            self.prepare_inputs(times, weather)
+            self.aoi_model()
+            self.spectral_model()
+            self.effective_irradiance_model()
+            self.temp_model()
         self.dc_model()
         self.ac_model()
         self.losses_model()
 
         return self
+
+
+def sdm_campanelli(mc):
```

Comment:
interesting idea to make a function here rather than a ModelChain method. It doesn't follow the pattern, but I could be convinced that all of the wrappers should look like this.

Reply:
Given the current architecture where the `PVSystem` wrappers are not ultimately responsible for determining the actual computations in a performance simulation, this is the "shortest path" I could come up with ;). It also pushes the idea that the "user" API could be the same as the "developer" API.